orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -61,28 +60,38 @@ GradientGraphBuilder::GradientGraphBuilder(Graph* graph,
     y_nodes_.insert(node);
   }
 
+  reachable_nodes_ = ReverseBFS(y_nodes_);
+
+  // building x_nodes_
   for (const auto& name : x_node_arg_names) {
     const NodeArg* node_arg = graph->GetNodeArg(name);
     if (!node_arg) {
       ORT_THROW("Node arg ", name, " is not found in the graph.");
     }
     x_node_args_.insert(node_arg);
 
-    vector<const Node*> nodes = graph_->GetConsumerNodes(name);
+    std::vector<const Node*> nodes = graph_->GetConsumerNodes(name);
     if (nodes.empty()) {
       ORT_THROW(name, " couldn't find the consumer node.");
     }
 
-    string grad_arg_name = GradientBuilderBase::GradientName(name);
-    pending_[grad_arg_name] = static_cast<int>(nodes.size());
+    std::string grad_arg_name = GradientBuilderBase::GradientName(name);
+    pending_[grad_arg_name] = 0;
 
-    x_nodes_.insert(nodes.begin(), nodes.end());
+    for (const Node* node : nodes) {
+      if (IsReachable(node)) {
+        pending_[grad_arg_name] += 1;
+        x_nodes_.insert(node);
+      } else {
+        LOGS(logger_, WARNING) << node->Name() << " is unreachable for gradient back propagation.";
```

Comment:
>LOGS(logger_, WARNING) [](start = 8, length = 22)

Would it better to collect all unreachable nodes and log outside of the loop?

Reply:
yes.